apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java

```diff
@@ -93,15 +93,29 @@ private Object parseScope(TenantOrDomainAndCredentials<T> credentials) {
       checkArgument(SCOPE_PREFIXES.contains(parts[0]), "Scope prefix should be: %s", SCOPE_PREFIXES);
 
       if (PROJECT.equals(parts[0])) {
-         Object domainScope = credentials.tenantOrDomainId() != null ? Id.create(credentials.tenantOrDomainId()) : Name
-               .create(credentials.tenantOrDomainName());
-         return ProjectScope.create(ProjectName.create(parts[1], domainScope));
+         return ProjectScope.create(ProjectName.create(parts[1], parseProjectDomain(credentials, true)));
       } else if (PROJECT_ID.equals(parts[0])) {
-         return ProjectIdScope.create(Id.create(parts[1]));
+         // tenant (name/id) was never used for project-id; so try to keep
+         // backward compatibility
+         return ProjectIdScope.create(ProjectId.create(parts[1], parseProjectDomain(credentials, false)));
       } else if (DOMAIN.equals(parts[0])) {
          return DomainScope.create(Name.create(parts[1]));
       } else {
          return DomainIdScope.create(Id.create(parts[1]));
       }
    }
+
+   private Object parseProjectDomain(TenantOrDomainAndCredentials<T> credentials, boolean useTenant) {
```

Comment:
Do we really need this boolean `useTenant` flag? Could we just change precedence and check the new properties first, then fallback to the tenant ones?

Reply:
For project/id, there is [currently](https://github.com/jclouds/jclouds/blob/master/apis/openstack-keystone/src/main/java/org/jclouds/openstack/keystone/v3/binders/BindAuthToJsonPayload.java#L100) no domain created on tenant properties => it was more for backwardcompatibility (use new properties only for project/id, both for project/name).